ompi/util/count_disp_array.h

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2024   Triad National Security, LLC. All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+#ifndef OMPI_UTIL_COUNT_DISP_ARRAY_H
+#define OMPI_UTIL_COUNT_DISP_ARRAY_H
+
+#include <stdlib.h>
+#include <stddef.h>
+#include <stdint.h>
+
+enum ompi_count_array_type {
+    OMPI_COUNT_ARRAY_TYPE_INT,
+    OMPI_COUNT_ARRAY_TYPE_SIZE_T,
+};
+
+typedef struct ompi_count_array {
+    enum ompi_count_array_type type;
+    union {
+        const int *int_array;
+        const size_t *size_t_array;
+    } data;
+} ompi_count_array;
+
+/* Initialize an int variant of the count array */
+static inline void ompi_count_array_init(ompi_count_array *array, const int *data)
+{
+    array->type = OMPI_COUNT_ARRAY_TYPE_INT;
+    array->data.int_array = data;
+}
+
+/* Initialize a bigcount variant of the count array */
+static inline void ompi_count_array_init_c(ompi_count_array *array, const size_t *data)
+{
+    array->type = OMPI_COUNT_ARRAY_TYPE_SIZE_T;
+    array->data.size_t_array = data;
+}
+
+#if OPAL_C_HAVE__GENERIC
+#define OMPI_COUNT_ARRAY_INIT(array, data) _Generic((data), \
+                                                    int *: ompi_count_array_init, \
+                                                    const int *: ompi_count_array_init, \
+                                                    size_t *: ompi_count_array_init_c, \
+                                                    const size_t *: ompi_count_array_init_c)(array, data)
+#else
+#define OMPI_COUNT_ARRAY_INIT(array, data) \
+    do { \
+        if (sizeof(*data) == sizeof(int)) { \
+            ompi_count_array_init(array, (const int *) data); \
+        } else if (sizeof(*data) == sizeof(size_t)) { \
+            ompi_count_array_init_c(array, (const size_t *) data); \
+        } \
+    } while (0)
```

Comment:
Parens around macro parameters:

```suggestion
#define OMPI_COUNT_ARRAY_INIT(array, data) \
    do { \
        if (sizeof(*(data)) == sizeof(int)) { \
            ompi_count_array_init((array), (const int *) (data)); \
        } else if (sizeof(*(data)) == sizeof(size_t)) { \
            ompi_count_array_init_c((array), (const size_t *) (data)); \
        } \
    } while (0)
```